wheatley/row_generation/helpers.py

```diff
@@ -74,3 +74,11 @@ def convert_to_bell_string(bell: int) -> str:
 def rounds(number_of_bells: int) -> Row:
     """ Generate rounds on the given number of bells. """
     return Row([Bell.from_number(i) for i in range(1, number_of_bells + 1)])
+
+
+def generate_starting_row(number_of_bells: int, start_row: str = None) -> Row:
+    """ Generate the starting row as rounds or a custom input. """
+    if start_row is None:
+        return rounds(number_of_bells)
+
+    return Row([Bell.from_str(i) for i in start_row])
```

Comment:
I'm not sure it has to contain every number once.
On init RowGenerator only knows about the stage and not the number of bells if I am reading it right.
Say you ring Doubles but want the 4 as cover, you might enter "12356" so checking all bells in the range 1-5 are used would break incorrectly.

I can check that the bells are unique and add any missing bells in to the end of the change.

The only oddity I have found doing that is if you ring doubles and enter a starting row of "4321" we will expand that to "432156" which is fine, but we are adding a working bell in, not sure if we want that or would rather error and force a user to enter it

Reply:
I agree with @centreboard here - we should verify that the input row is valid along with the other args.  Currently you could input `--start-row 414` and nothing would break (except that you'd get shedloads of 'bell on wrong stroke' errors because the 4 is rung twice in every row).  Also, I suspect that in the vast majority of cases invalid rows are simply a mistake on behalf of the user so it would be nice to throw an upfront error (anyway rows have to contain each bell exactly once in order to be classified as change ringing).  I implemented checks like this for a separate project and it works nicely - [here](https://github.com/kneasle/3rd-yr-proj/blob/master/core/src/row.rs#L407-L430)'s the code in that project, which might be helpful (it's in Rust though which is not very widely used).

So, what I suggest is that the incoming `--start-row` arg is parsed and verified along with the other args (e.g. call strings, peal speed, etc.) - this should probably add another function to `parsing.py` with appropriate error classes for missing and duplicated bells.  We don't verify the stage at this point, because that could change whilst Wheatley is running.  Then we pass the parsed and verified row into `Bot.__init__` and then everything else is already done.

Reply:
@chrisjfield is right we could just append missing bells onto the end. My suspicion is entering `13256` for doubles is more likely to be accidentally not typing `4` than wanting it as a cover bell. If Wheatley adds it, there should probably be an information or warning logged.
If we expand to add cover bell, I'm also fine with expanding to add working bells (so `1423` could work for minor).

The short form I'd probably use most want to use is `14325` for doubles on 5, 6 or more bells.

Ultimately these are really edge cases and I'm happy with whatever's cleanest to code such that Wheatley ends up with valid rows, and we've communicated validation errors back to users.

Reply:
Hmmm reading my comments I haven't been very clear... so my proposed idea is the following:
- The user enters some string (say `--start-row 1423`)
- Wheatley parses this string into a row of the smallest possible stage while still in the `main` function, in this case `[1, 4, 2, 3]`.  If this row is invalid, then we should explain the error somehow and terminate, e.g.:
  - For `1442` we should emit something like `Bell '4' is duplicated`
  - For `124` we either emit `Bell '3' is missing` or `Bell '4' is too big`.  Both are equally valid but the former is probably more intuitive.
- This small but valid row is then passed into `Bot` as the starting row, and has consecutive bells appended to the end to extend it to the correct number of bels (just like how we currently add cover bells).  This process can't invalidate a row, so it's fine to do without worries about misinterpretation.  If the unextended row already has too many bells, then we emit a warning and crash on entering changes (similar to ringing e.g. major on 6 bells).

This way, I think we get the best of all worlds - Wheatley helpfully reports potential errors and can't ring any invalid rows, but we also get the nice ability to permute the little bells without typing all the rounds.  It also has the nice side effect that `""` parses to the empty row `[]` which then expands to rounds on any stage, so we don't have to handle rounds as a special case.

Reply:
I was playing with this last night and have something locally that works (I think) but was waiting to see what your thoughts were. It seems similar to what you suggest above but with a slight difference so interested in what you think.

What I have at the moment is:
* The user enters some string (say --start-row 1423)
* On init we parse the string using the stage
** if duplicate bells are received we throw a value error
** if bells are missing we append to the end in order (so -- start-row 134 with stage 5 becomes 13425)
* In bot we re-generate using the actual number of bells in the tower, so if there are 6 bells it now becomes 134256
* Before we go into changes we do some extra checks:
** check the inputted starting row is equal or shorter to the number of bells in the tower (just to allow nice messaging)
** check the actual starting row equals the number of bells in the tower.

I think this is similar but allows for inputting shorter starting rows where you want unusual cover bells. This is obviously an edge edge case but may as well let it go IMO.

It also indirectly catches out of range errors as they would end up with a starting row longer than the number of bells.
I think it's best to check here as the row might not start valid but could become valid when the tower changes size etc so seems odd to error before they have a chance to fix it.

The only oddity I found is it means Wheatly can add bells which are working bells. As @centreboard mentioned that might be ok. If we don't like it it's easy enough to add another validation that the inputted row is >= to the stage in length.

Hope that makes sense, thoughts welcome!